chalk-solve/src/clauses/syntactic_eq.rs

```diff
@@ -0,0 +1,165 @@
+use std::{iter, mem::replace};
+
+use chalk_engine::fallible::Fallible;
+use chalk_ir::{
+    cast::Cast,
+    fold::{shift::Shift, Fold, Folder, SuperFold},
+    interner::Interner,
+    AliasEq, AliasTy, Binders, BoundVar, DebruijnIndex, Goal, GoalData, Goals, ParameterKind,
+    ParameterKinds, ProgramClause, ProgramClauseData, ProgramClauseImplication, QuantifierKind, Ty,
+    TyData,
+};
+
+pub fn syn_eq_lower<I: Interner, T: Fold<I>>(interner: &I, clause: &T) -> <T as Fold<I>>::Result {
+    let mut folder = SynEqFolder {
+        interner,
+        new_params: vec![],
+        new_goals: vec![],
+        binders_len: 0,
+    };
+
+    clause
+        .fold_with(&mut folder, DebruijnIndex::INNERMOST)
+        .unwrap()
+}
+
+struct SynEqFolder<'i, I: Interner> {
+    interner: &'i I,
+    new_params: Vec<ParameterKind<()>>,
+    new_goals: Vec<Goal<I>>,
+    binders_len: usize,
+}
+
+impl<'i, I: Interner> Folder<'i, I> for SynEqFolder<'i, I> {
+    fn as_dyn(&mut self) -> &mut dyn Folder<'i, I> {
+        self
+    }
+
+    fn fold_ty(&mut self, ty: &Ty<I>, outer_binder: DebruijnIndex) -> Fallible<Ty<I>> {
+        let interner = self.interner;
+        let bound_var = BoundVar::new(DebruijnIndex::INNERMOST, self.binders_len);
+
+        let new_ty = TyData::BoundVar(bound_var).intern(interner);
+        match ty.data(interner) {
+            TyData::Alias(alias @ AliasTy::Projection(_)) => {
+                self.new_params.push(ParameterKind::Ty(()));
+                self.new_goals.push(
+                    AliasEq {
+                        alias: alias.clone(),
+                        ty: new_ty.clone(),
+                    }
+                    .cast(interner),
+                );
+                self.binders_len += 1;
+                ty.super_fold_with(self, outer_binder)?;
+                Ok(new_ty)
+            }
+            TyData::Function(_) => Ok(ty.clone()),
+            _ => Ok(ty.super_fold_with(self, outer_binder)?),
+        }
+    }
+
+    fn fold_program_clause(
```

Comment:
I would include a comment here that gives the overall strategy, which the function *body* can reference. Of course this was outlined in great detail in the module comment, so something like this would work. In these cases, I like to give very concrete example so that, in the code comments, I can refer to very specific things. e.g.,

```
/// Convert a program clause to rem
///
/// Consider this (nonsense) example:
///
/// ```notrust
/// forall<X> { 
///     Implemented(<X as Iterator>::Item>: Debug) :- 
///         Implemented(X: Debug)
/// }
/// ```
///
/// we would lower this into:
///
/// ```notrust
/// forall<X, Y> { 
///     Implemented(Y: Debug) :- 
///         AliasEq(<X as Iterator>::Item>, Y),
///         Implemented(X: Debug)
/// }
/// ```
```